### PR TITLE
INT-28 Create skill api

### PR DESCRIPTION
**Pull request by Born2Flex**

### CODE REVIEW

File: src/test/java/ua/edu/internship/interview/service/business/SkillServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import ua.edu.internship.interview.data.repository.SkillRepository;
+import ua.edu.internship.interview.service.dto.SkillTreeDto;
+import ua.edu.internship.interview.service.utils.exceptions.NoSuchEntityException;
+
+import java.util.ArrayList;
```
Comment (Born2Flex) [inline src/test/java/ua/edu/internship/interview/service/business/SkillServiceTest.java:12]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.ArrayList'

File: src/test/java/ua/edu/internship/interview/service/business/SkillServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.List;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (Born2Flex) [inline src/test/java/ua/edu/internship/interview/service/business/SkillServiceTest.java:16]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'org.junit.jupiter.api.Assertions.assertEquals'

File: src/test/java/ua/edu/internship/interview/service/business/SkillServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+
+@ExtendWith(MockitoExtension.class)
```
Comment (Born2Flex) [inline src/test/java/ua/edu/internship/interview/service/business/SkillServiceTest.java:23]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.whitespace.EmptyLineSeparatorCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>'CLASS_DEF' has more than 1 empty lines before.

File: src/main/java/ua/edu/internship/interview/data/repository/SkillRepository.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
+package ua.edu.internship.interview.data.repository;
+
+import org.springframework.data.mongodb.repository.Aggregation;
+import org.springframework.data.mongodb.repository.MongoRepository;
+import ua.edu.internship.interview.data.documents.SkillDocument;
+import ua.edu.internship.interview.service.dto.SkillTreeDto;
+
+import java.util.List;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/data/repository/SkillRepository.java:8]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.List'

File: src/main/java/ua/edu/internship/interview/service/business/SkillService.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package ua.edu.internship.interview.service.business;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+import ua.edu.internship.interview.data.repository.SkillRepository;
+import ua.edu.internship.interview.service.dto.SkillTreeDto;
+import ua.edu.internship.interview.service.utils.exceptions.NoSuchEntityException;
+
+import java.util.List;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/service/business/SkillService.java:10]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.List'

File: src/main/java/ua/edu/internship/interview/service/dto/SkillTreeDto.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package ua.edu.internship.interview.service.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+
+import java.util.ArrayList;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/service/dto/SkillTreeDto.java:8]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.ArrayList'

File: src/main/java/ua/edu/internship/interview/config/migrations/SkillsInitializerChangeUnit.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+import io.mongock.api.annotations.ChangeUnit;
+import io.mongock.api.annotations.Execution;
+import io.mongock.api.annotations.RollbackExecution;
+import lombok.RequiredArgsConstructor;
+import org.bson.types.ObjectId;
+import ua.edu.internship.interview.data.documents.SkillDocument;
+import ua.edu.internship.interview.data.repository.SkillRepository;
+
+import java.util.List;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/config/migrations/SkillsInitializerChangeUnit.java:11]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.List'

File: src/main/java/ua/edu/internship/interview/config/SwaggerConfig.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package ua.edu.internship.interview.config;
+
+import io.swagger.v3.oas.models.OpenAPI;
+import io.swagger.v3.oas.models.info.Info;
+import io.swagger.v3.oas.models.servers.Server;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+import java.util.List;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/config/SwaggerConfig.java:10]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.List'

File: src/main/java/ua/edu/internship/interview/config/SwaggerConfig.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+@Configuration
+public class SwaggerConfig {
+    @Value("${project.version}")
+    private String version;
+    @Value("${domain.back}")
+    private String devServer;
+
+    @Bean
+    public OpenAPI openAPI() {
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/config/SwaggerConfig.java:20]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.naming.AbbreviationAsWordInNameCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Abbreviation in name 'openAPI' must contain no more than '2' consecutive capital letters.

File: src/main/java/ua/edu/internship/interview/web/controller/SkillController.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import io.swagger.v3.oas.annotations.tags.Tag;
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import ua.edu.internship.interview.service.business.SkillService;
+import ua.edu.internship.interview.service.dto.SkillTreeDto;
+
+import java.util.List;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/web/controller/SkillController.java:12]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.List'

File: src/main/java/ua/edu/internship/interview/web/handler/GlobalExceptionHandler.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.FieldError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import ua.edu.internship.interview.service.utils.exceptions.BaseException;
+
+import java.util.HashMap;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/web/handler/GlobalExceptionHandler.java:14]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.util.HashMap'

File: src/main/java/ua/edu/internship/interview/web/handler/ErrorResponse.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package ua.edu.internship.interview.web.handler;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import org.springframework.http.HttpStatus;
+
+import java.time.LocalDateTime;
```
Comment (Born2Flex) [inline src/main/java/ua/edu/internship/interview/web/handler/ErrorResponse.java:8]: 🚫 **[checkstyle]** <com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Extra separation in import group before 'java.time.LocalDateTime'

**Comment by Born2Flex** [comment]
### JaCoCo Report
|Overall Project|100%|✔️|
|:-|:-|:-:|
|Files changed|100%|✔️|
<br>

|File|Coverage||
|:-|:-|:-:|
|[SkillService.java](https://github.com/Born2Flex/interview/blob/ba98912a5f9452d1a978d00687d050b5007528d8/src%2Fmain%2Fjava%2Fua%2Fedu%2Finternship%2Finterview%2Fservice%2Fbusiness%2FSkillService.java)|100%|✔️|

File: README.md [hunk 13]
```diff
@@ -18,3 +18,32 @@ The documentation is divided into several sections:
 - [Business Cases](docs/business_cases.md)
 - [Architecture diagram](docs/architecture_diagram.md)
 - [ER Diagram](docs/er_diagram.md)
+
```
Comment (proyalko) [inline README.md:21]: The readme file should contain technical descriptions for other developers: how to start the application locally, how to use it, and what you need to install and run before starting the application.

Comment (proyalko) [reply README.md:21]: Discussed

File: .github/workflows/pipeline.yml [hunk 14]
```diff
@@ -38,6 +38,7 @@ jobs:
         env:
           SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
           SONAR_HOST_URL: 'https://sonarcloud.io/'
+          continue-on-error: 'true'
```
Comment (proyalko) [inline .github/workflows/pipeline.yml:41]: Why do we need to continue on error for Jacoco and Sonar?

File: README.md [hunk 15]
```diff
@@ -18,3 +18,32 @@ The documentation is divided into several sections:
...
+## Run locally
```
Comment (proyalko) [inline README.md:22]: Please add the docker-compose config for Mongo to the repo

Comment (proyalko) [reply README.md:22]: Discussed

File: src/main/java/ua/edu/internship/interview/service/dto/skill/SkillDto.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package ua.edu.internship.interview.service.dto.skill;
+
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+@Data
+@EqualsAndHashCode(callSuper = true)
+public class SkillDto extends SkillBaseDto {
```
Comment (proyalko) [inline src/main/java/ua/edu/internship/interview/service/dto/skill/SkillDto.java:8]: Why do we need this SkillDto? I haven't found any usage of it

File: src/main/java/ua/edu/internship/interview/service/dto/skill/SkillBaseDto.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
+public class SkillBaseDto {
```
Comment (proyalko) [inline src/main/java/ua/edu/internship/interview/service/dto/skill/SkillBaseDto.java:12]: Why do we need SkillBaseDto?

File: .github/workflows/pipeline.yml (hunk 14 above)
Comment (Born2Flex) [reply .github/workflows/pipeline.yml:41]: I think it would be better to continue pipeline and see all errors

Comment (Born2Flex) [reply .github/workflows/pipeline.yml:41]: Removed continue on error in jacoco step

File: src/main/java/ua/edu/internship/interview/service/dto/skill/SkillDto.java (hunk 16 above)
Comment (Born2Flex) [reply src/main/java/ua/edu/internship/interview/service/dto/skill/SkillDto.java:8]: This DTO is needed in next pull request. Discussed

File: src/main/java/ua/edu/internship/interview/service/dto/skill/SkillBaseDto.java (hunk 17 above)
Comment (Born2Flex) [reply src/main/java/ua/edu/internship/interview/service/dto/skill/SkillBaseDto.java:12]: Discussed

File: src/test/java/ua/edu/internship/interview/service/business/SkillServiceIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+    @Test
+    void testGetSkillTreeByIdNotFound() {
+        String invalidId = new ObjectId().toHexString();
+
+        assertThrows(NoSuchEntityException.class, () -> skillService.getSkillTreeById(invalidId));
+    }
+
+    @Test
+    void testGetSkillTreeByIdFound() {
+        ObjectId id = new ObjectId();
```
Comment (proyalko) [inline src/test/java/ua/edu/internship/interview/service/business/SkillServiceIntegrationTest.java:57]: Try to use an external query to init/cleanup mongo data before and after the test to be repository independent in this case

Comment (proyalko) [reply src/test/java/ua/edu/internship/interview/service/business/SkillServiceIntegrationTest.java]: Discussed, fixed

**Pull request merged**